examples/common/json/JsonTLV.cpp

```diff
@@ -0,0 +1,311 @@
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements interfaces for debugging and logging
+ *      CHIP TLV.
+ *
+ */
+
+#ifndef __STDC_FORMAT_MACROS
+#define __STDC_FORMAT_MACROS
+#endif
+
+#include <ctype.h>
+#include <inttypes.h>
+#include <string.h>
+
+#include <lib/core/TLV.h>
+#include <lib/core/TLVUtilities.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+#include "JsonTLV.h"
+
+namespace chip {
+
+namespace TLV {
+
+namespace Json {
+
+void ENFORCE_FORMAT(1, 2) TLVDumpWriter(const char * aFormat, ...)
+{
+    va_list args;
+
+    va_start(args, aFormat);
+
+    vprintf(aFormat, args);
+
+    va_end(args);
+}
+
+static void ENFORCE_FORMAT(2, 3) FormattedWrite(Encoding::BufferWriter & writer, const char * aFormat, ...)
+{
+    char out_buf[100];
+
+    va_list args;
+
+    va_start(args, aFormat);
+
+    vsnprintf(out_buf, sizeof(out_buf), aFormat, args);
+
+    va_end(args);
+
+    writer.Put(out_buf);
+}
+
+/**
+ *  Write the TLV element referenced by @a aReader in human-readable form using
+ *  @a aWriter.
+ *
+ *  @param[in]     aWriter   The writer to log the TLV data.
+ *  @param[in]     aIndent   The indentation for logging the current depth into
+ *                           the TLV data.
+ *  @param[in]     aReader   A read-only reference to the TLV reader containing
+ *                           the TLV data to log.
+ *  @param[in]     aDepth    The current depth into the TLV data.
+ *
+ */
+static void WriteHandlerJSON(Encoding::BufferWriter & aWriter, const char * aIndent, const TLVReader & aReader, size_t aDepth)
+{
+    const TLVType type     = aReader.GetType();
+    const Tag tag          = aReader.GetTag();
+    const uint32_t len     = aReader.GetLength();
+    const uint8_t * strbuf = nullptr;
+    CHIP_ERROR err         = CHIP_NO_ERROR;
+    TLVReader temp;
+    TLVTagControl tagControl;
+
+    temp.Init(aReader);
+    tagControl = static_cast<TLVTagControl>(temp.GetControlByte() & kTLVTagControlMask);
+
+    for (size_t i = 0; i < aDepth; i++)
+        FormattedWrite(aWriter, "%s", aIndent);
+
+    if (TLVTypeIsContainer(type))
+    {
+        if (tagControl == TLVTagControl::ContextSpecific)
+        {
+            FormattedWrite(aWriter, "\"0x%x\" : ", TagNumFromTag(tag));
+        }
+
+        // add any prefix
+        switch (type)
+        {
+        case kTLVType_Structure:
+            FormattedWrite(aWriter, "{\n");
+            break;
+
+        case kTLVType_Array:
+            FormattedWrite(aWriter, "[\n");
+            break;
+
+        case kTLVType_List:
+            FormattedWrite(aWriter, "[\n");
+            break;
+
+        default:
+            break;
+        }
+    }
+    else
+    {
+        if (tagControl == TLVTagControl::ContextSpecific)
```

Comment:
Again, this does not seem to handle profile tags right.  But also, shouldn't that `"tag" : ` prefix just be written out independently of whether the tag is a container or not?